src/tcp_sock.rs

```diff
@@ -324,14 +354,14 @@ impl Drop for Inner {
 }
 
 /// Serialize given value and write to a buffer with 4 byte length header.
-fn serialize_with_len<T: Serialize>(value: T) -> ::Res<Vec<u8>> {
-    let mut data = Cursor::new(Vec::with_capacity(mem::size_of::<u32>()));
+fn serialize_with_len<T: Serialize>(value: T, enc_ctx: &EncryptContext) -> ::Res<Vec<u8>> {
+    let encrypted = enc_ctx.encrypt(&value)?;
 
-    let _ = data.write_u32::<LittleEndian>(0);
-    serialise_into(&value, &mut data)?;
-    let len = data.position() - mem::size_of::<u32>() as u64;
-    data.set_position(0);
-    data.write_u32::<LittleEndian>(len as u32)?;
+    let mut data = Cursor::new(Vec::with_capacity(mem::size_of::<u32>() + encrypted.len()));
+    let _ = data.write_u32::<LittleEndian>(encrypted.len() as u32);
+    // TODO(povilas): if safe_crypto implements encrypt_into, use that to reduce data copying
+    let bytes_written = data.write(&encrypted)?;
+    debug_assert_eq!(bytes_written, encrypted.len());
```

Comment:
can we not use these pls as CI runs tests only in release mode so this is not really enforceable ?

Reply:
Well I expect this to always be true, since `Cursor` has a vec with proper length, but this is a safeguard just in case my expectations are wrong. I don't want to make a runtime check since I expect it to always be true, and that's what `debug_assert` macros do. If CI won't catch it, testiing on localhost would. Alternatively, I could ignore the `data.write()` result, but that wouldn't save us anything :)